configfile: 'config.json'

rule all:
    input:
        '../original_data/H1_Meso_D3_FACS_RNA_S2_I1_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_I2_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_R1_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_R2_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_CB_UMI.fastq.gz',
        'downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cdna.fq.gz',
        'downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cb_umi.fq.gz',
        'outs/Solo.out/Gene/filtered/barcodes.tsv',
        'outs/Solo.out/Gene/filtered/features.tsv',
        'outs/Solo.out/Gene/filtered/matrix.mtx',
        'outs/Aligned.sortedByCoord.out.bam',
        'outs/Solo.out/Gene/filtered/metrics.csv'

rule downloadFromEBI:
    input:
        '../fq_url.txt'
    output:
        '../original_data/H1_Meso_D3_FACS_RNA_S2_I1_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_I2_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_R1_001.fastq.gz',
        '../original_data/H1_Meso_D3_FACS_RNA_S2_R2_001.fastq.gz'
    shell:
        ''' wget -nc -P ../original_data -i {input}
        '''

rule prepareCBUMI:
    input:
        i1='../original_data/H1_Meso_D3_FACS_RNA_S2_I1_001.fastq.gz',
        i2='../original_data/H1_Meso_D3_FACS_RNA_S2_I2_001.fastq.gz',
        umi='../original_data/H1_Meso_D3_FACS_RNA_S2_R1_001.fastq.gz'
    output:
        '../original_data/H1_Meso_D3_FACS_RNA_S2_CB_UMI.fastq.gz'
    shell:
        ''' ../../../scripts/collect_cb_umi_issaac_plate.py \
            {input.i1} {input.i2} {input.umi} | \
            gzip > {output}
        '''

rule subsample:
    input:
        cdna='../original_data/H1_Meso_D3_FACS_RNA_S2_R2_001.fastq.gz',
        cb='../original_data/H1_Meso_D3_FACS_RNA_S2_CB_UMI.fastq.gz'
    output:
        cdna='downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cdna.fq.gz',
        cb='downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cb_umi.fq.gz'
    shell:
        ''' seqtk sample -s 42 {input.cdna} 0.14 | gzip > {output.cdna}
            seqtk sample -s 42 {input.cb} 0.14 | gzip > {output.cb}
        '''

rule star:
    input:
        pg=config['star_bin'],
        cdna='downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cdna.fq.gz',
        cb='downsampled_fq/H1_Meso_D3_FACS_RNA_ds_cb_umi.fq.gz',
        wl='../H1_Meso_D3_FACS_RNA_whitelist.txt'
    output:
        'outs/Solo.out/Gene/filtered/barcodes.tsv',
        'outs/Solo.out/Gene/filtered/features.tsv',
        'outs/Solo.out/Gene/filtered/matrix.mtx',
        'outs/Aligned.sortedByCoord.out.bam'
    params:
        idx=config['star_idx']
    threads: 40
    shell:
        ''' {input.pg} --genomeDir {params.idx} --outFileNamePrefix ./outs/ \
            --readFilesCommand zcat --readFilesIn {input.cdna} {input.cb} \
            --soloCBstart 1 --soloCBlen 16 --soloUMIstart 17 --soloUMIlen 10 \
            --soloType CB_UMI_Simple --clip3pNbases 3 \
            --soloCBwhitelist {input.wl} \
            --soloCellFilter EmptyDrops_CR --runThreadN {threads} \
            --outSAMattributes CB UB --outSAMtype BAM SortedByCoordinate
        '''

rule getMetrics:
    input:
        bc='outs/Solo.out/Gene/filtered/barcodes.tsv',
        mtx='outs/Solo.out/Gene/filtered/matrix.mtx'
    output:
        'outs/Solo.out/Gene/filtered/metrics.csv'
    shell:
        ''' ../../../scripts/output_qc_mtx {input.mtx} {input.bc} {output}
        '''
