configfile: 'config.json'

rule all:
    input:
        'raw_data/SRR8528318.sra',
        'raw_data/SRR8528318_1.fastq.gz',
        'raw_data/SRR8528318_2.fastq.gz',
        'outs/Solo.out/Gene/filtered/barcodes.tsv',
        'outs/Solo.out/Gene/filtered/features.tsv',
        'outs/Solo.out/Gene/filtered/matrix.mtx',
        'outs/Aligned.sortedByCoord.out.bam',
        'outs/Solo.out/Gene/filtered/metrics.csv'

rule downloadSra:
    output:
        'raw_data/SRR8528318.sra'
    shell:
        ''' prefetch SRR8528318 -o raw_data/SRR8528318.sra
        '''

rule dumpFastq:
    input:
        'raw_data/SRR8528318.sra'
    output:
        'raw_data/SRR8528318_1.fastq.gz',
        'raw_data/SRR8528318_2.fastq.gz'
    threads: 8
    shell:
        ''' fasterq-dump --split-3 -e {threads} -O raw_data {input}
            gzip raw_data/SRR8528318_1.fastq
            gzip raw_data/SRR8528318_2.fastq
        '''

rule star:
    input:
        pg=config['star_bin'],
        cdna='raw_data/SRR8528318_2.fastq.gz',
        cb='raw_data/SRR8528318_1.fastq.gz'
    output:
        'outs/Solo.out/Gene/filtered/barcodes.tsv',
        'outs/Solo.out/Gene/filtered/features.tsv',
        'outs/Solo.out/Gene/filtered/matrix.mtx',
        'outs/Aligned.sortedByCoord.out.bam'
    params:
        idx=config['star_idx']
    threads: 40
    shell:
        ''' {input.pg} --genomeDir {params.idx} --outFileNamePrefix ./outs/ \
            --readFilesCommand zcat --readFilesIn {input.cdna} {input.cb} \
            --soloCBstart 1 --soloCBlen 12 --soloUMIstart 13 --soloUMIlen 8 \
            --soloBarcodeReadLength 0 \
            --soloType CB_UMI_Simple --clip3pNbases 45 \
            --soloCBwhitelist None \
            --soloCellFilter EmptyDrops_CR --runThreadN {threads} \
            --outSAMattributes CB UB --outSAMtype BAM SortedByCoordinate
        '''

rule getMetrics:
    input:
        bc='outs/Solo.out/Gene/filtered/barcodes.tsv',
        mtx='outs/Solo.out/Gene/filtered/matrix.mtx'
    output:
        'outs/Solo.out/Gene/filtered/metrics.csv'
    shell:
        ''' ../../../scripts/output_qc_mtx {input.mtx} {input.bc} {output}
        '''
